### PR TITLE
feat: check CoreDNS

### PR DESCRIPTION
**Pull request by c3y1huang**
#### Which issue(s) this PR fixes:

Issue longhorn/longhorn#9752

#### What this PR does / why we need it:

Check if CoreDNS has more than 2 ready replicas.

#### Special notes for your reviewer:

`None`

#### Additional documentation or context

`None`

### CODE REVIEW

**Comment by c3y1huang** [comment]
@coderabbitai review

**Comment by c3y1huang** [comment]
Test results:
1. CoreDNS with 1 replica.
   ```
   > ./bin/longhornctl-linux-amd64 check preflight --image="c3y1huang/research:longhorn-cli-443"
   INFO[2024-11-14T10:42:39+08:00] Initializing preflight checker               
   INFO[2024-11-14T10:42:39+08:00] Cleaning up preflight checker                
   INFO[2024-11-14T10:42:39+08:00] Running preflight checker                    
   INFO[2024-11-14T10:42:43+08:00] Retrieved preflight checker result:
   ip-10-0-2-6:
     info:
     - Service iscsid is running
     - NFS4 is supported
     - Package nfs-client is installed
     - Package open-iscsi is installed
     - Package cryptsetup is installed
     - Module dm_crypt is loaded
     warn:
     - CoreDNS is set with fewer than 2 replicas; consider increasing replica count for high availability
   ip-10-0-2-21:
     info:
     - Service iscsid is running
     - NFS4 is supported
     - Package nfs-client is installed
     - Package open-iscsi is installed
     - Package cryptsetup is installed
     - Module dm_crypt is loaded
     warn:
     - CoreDNS is set with fewer than 2 replicas; consider increasing replica count for high availability
   ip-10-0-2-187:
     info:
     - Service iscsid is running
     - NFS4 is supported
     - Package nfs-client is installed
     - Package open-iscsi is installed
     - Package cryptsetup is installed
     - Module dm_crypt is loaded
     warn:
     - CoreDNS is set with fewer than 2 replicas; consider increasing replica count for high availability 
   INFO[2024-11-14T10:42:43+08:00] Cleaning up preflight checker                
   INFO[2024-11-14T10:42:43+08:00] Completed preflight checker   
   ```
1. CoreDNS with 2 replicas.
   ```
   > ./bin/longhornctl-linux-amd64 check preflight --image="c3y1huang/research:longhorn-cli-443"
   INFO[2024-11-14T10:43:45+08:00] Initializing preflight checker               
   INFO[2024-11-14T10:43:45+08:00] Cleaning up preflight checker                
   INFO[2024-11-14T10:43:46+08:00] Running preflight checker                    
   INFO[2024-11-14T10:43:49+08:00] Retrieved preflight checker result:
   ip-10-0-2-6:
     info:
     - CoreDNS is set with 2 replicas and 2 ready replicas
     - Service iscsid is running
     - NFS4 is supported
     - Package nfs-client is installed
     - Package open-iscsi is installed
     - Package cryptsetup is installed
     - Module dm_crypt is loaded
   ip-10-0-2-21:
     info:
     - CoreDNS is set with 2 replicas and 2 ready replicas
     - Service iscsid is running
     - NFS4 is supported
     - Package nfs-client is installed
     - Package open-iscsi is installed
     - Package cryptsetup is installed
     - Module dm_crypt is loaded
   ip-10-0-2-187:
     info:
     - CoreDNS is set with 2 replicas and 2 ready replicas
     - Service iscsid is running
     - NFS4 is supported
     - Package nfs-client is installed
     - Package open-iscsi is installed
     - Package cryptsetup is installed
     - Module dm_crypt is loaded 
   INFO[2024-11-14T10:43:49+08:00] Cleaning up preflight checker                
   INFO[2024-11-14T10:43:50+08:00] Completed preflight checker 
   ```

**Comment by c3y1huang** [comment]
@coderabbitai review

File: go.mod [hunk 1]
```diff
@@ -4,6 +4,8 @@ go 1.23
 
 toolchain go1.23.3
 
+replace github.com/longhorn/go-common-libs => github.com/c3y1huang/go-common-libs v0.0.0-20241114031240-e1117130cd3b
```
Comment (c3y1huang) [inline go.mod:7]: TODO: update after dependent vendor PR merged.

**Comment by c3y1huang** [comment]
> Hi @c3y1huang I have a question: should we change the number of replicas for the `CoreDNS` deployment in the cluster after running the install preflight? One concern I have is that the default number of replicas for the `CoreDNS` deployment is 1, but the user's original cluster might have been changed to 3. However, after running Longhorn cli, it might become 2, which could lead to issues.

This PR is proposing to add the CoreDNS check to the `longhornctl check preflight`. It doesn't change the CoreDNS setup as that responsibility should remain with the user, since CoreDNS is not managed by Longhorn.

**Review by shuo-wu — DISMISSED** [review]
LGTM

File: pkg/local/preflight/checker.go [hunk 2]
```diff
@@ -466,3 +468,36 @@ func (local *Checker) checkNFSv4Support() error {
...
+// replicas is less than 2, it logs a warning indicating that CoreDNS is
+// not set to run in multiple replicas. Additionally, it checks the number
+// of ready replicas in the deployment status and logs a warning if there
+// are fewer than 2 ready replicas.
+//
+// https://github.com/longhorn/longhorn/issues/9752
+func (local *Checker) checkCoreDNS() {
+	logrus.Info("Checking if CoreDNS has multiple replicas")
+
+	deployment, err := commonkube.GetDeployment(local.kubeClient, metav1.NamespaceSystem, consts.AppNameCoreDNS)
```
Comment (derekbit) [inline pkg/local/preflight/checker.go:486]: `AppNameCoreDNS` is `coredns`. If the k8s cluster is bootstrapped by rke2, the coredns deployment is `rke2-coredns-rke2-coredns`, is the check still valid?

File: go.mod (hunk 1 above)
Comment (derekbit) [reply go.mod:7]: dep is merged.

File: go.mod [hunk 3]
```diff
@@ -4,6 +4,8 @@ go 1.23
 
 toolchain go1.23.3
 
+replace github.com/longhorn/go-common-libs => github.com/c3y1huang/go-common-libs v0.0.0-20241128011818-831be76ab726
```
Comment (c3y1huang) [inline go.mod:7]: TODO: update after dependent vendor PR merged.

**Comment by c3y1huang** [comment]
Test Results:
- RKE2:
  ```
  INFO[2024-11-28T09:34:33+08:00] Initializing preflight checker               
  INFO[2024-11-28T09:34:33+08:00] Cleaning up preflight checker                
  INFO[2024-11-28T09:34:33+08:00] Running preflight checker                    
  INFO[2024-11-28T09:34:45+08:00] Retrieved preflight checker result:
  ip-10-0-2-111:
    info:
    - Kube DNS "rke2-coredns-rke2-coredns" is set with 2 replicas and 2 ready replicas
    - Service iscsid is running
    - NFS4 is supported
    - Package nfs-client is installed
    - Package open-iscsi is installed
    - Package cryptsetup is installed
    - Module dm_crypt is loaded
  ip-10-0-2-112:
    info:
    - Kube DNS "rke2-coredns-rke2-coredns" is set with 2 replicas and 2 ready replicas
    - Service iscsid is running
    - NFS4 is supported
    - Package nfs-client is installed
    - Package open-iscsi is installed
    - Package cryptsetup is installed
    - Module dm_crypt is loaded
  ip-10-0-2-119:
    info:
    - Kube DNS "rke2-coredns-rke2-coredns" is set with 2 replicas and 2 ready replicas
    - Service iscsid is running
    - NFS4 is supported
    - Package nfs-client is installed
    - Package open-iscsi is installed
    - Package cryptsetup is installed
    - Module dm_crypt is loaded 
  INFO[2024-11-28T09:34:45+08:00] Cleaning up preflight checker                
  INFO[2024-11-28T09:34:45+08:00] Completed preflight checker
  ```
- K3s:
  ```
  INFO[2024-11-28T09:49:13+08:00] Initializing preflight checker               
  INFO[2024-11-28T09:49:13+08:00] Cleaning up preflight checker                
  INFO[2024-11-28T09:49:13+08:00] Running preflight checker                    
  INFO[2024-11-28T09:49:17+08:00] Retrieved preflight checker result:
  ip-10-0-2-42:
    error:
    - 'Module dm_crypt is not loaded: failed to execute: nsenter [--mount=/host/proc/10437/ns/mnt --net=/host/proc/10437/ns/net grep dm_crypt /proc/modules], output , stderr : exit status 1'
    info:
    - Kube DNS "coredns" is set with 2 replicas and 2 ready replicas
    - Service iscsid is running
    - NFS4 is supported
    - Package nfs-client is installed
    - Package open-iscsi is installed
    - Package cryptsetup is installed
  ip-10-0-2-182:
    error:
    - 'Module dm_crypt is not loaded: failed to execute: nsenter [--mount=/host/proc/10201/ns/mnt --net=/host/proc/10201/ns/net grep dm_crypt /proc/modules], output , stderr : exit status 1'
    info:
    - Kube DNS "coredns" is set with 2 replicas and 2 ready replicas
    - Service iscsid is running
    - NFS4 is supported
    - Package nfs-client is installed
    - Package open-iscsi is installed
    - Package cryptsetup is installed
  ip-10-0-2-228:
    error:
    - 'Module dm_crypt is not loaded: failed to execute: nsenter [--mount=/host/proc/11666/ns/mnt --net=/host/proc/11666/ns/net grep dm_crypt /proc/modules], output , stderr : exit status 1'
    info:
    - Kube DNS "coredns" is set with 2 replicas and 2 ready replicas
    - Service iscsid is running
    - NFS4 is supported
    - Package nfs-client is installed
    - Package open-iscsi is installed
    - Package cryptsetup is installed 
  INFO[2024-11-28T09:49:17+08:00] Cleaning up preflight checker                
  INFO[2024-11-28T09:49:17+08:00] Completed preflight checker
  ```

File: pkg/local/preflight/checker.go (hunk 2 above)
Comment (c3y1huang) [reply pkg/local/preflight/checker.go]: Good catch. Thank you and updated.

**Review by derekbit — APPROVED** [review]
LGTM. Thank you @c3y1huang

**Comment by c3y1huang** [comment]
@mergify backport v1.7.x

**Pull request merged**